### ISSUE TITLE
Wrong winner reported after a white resignation

### ISSUE THREAD

**Issue by roy7** [gcp/leela-zero#338]
As reported here https://github.com/gcp/leela-zero/issues/78#issuecomment-350568901, White resigned but game score shows W+Resign with white as winner.

I'm guessing the problem is in this spot:

>     } else if (m_moveDone.compare(QStringLiteral("resign"),
>                                  Qt::CaseInsensitive) == 0) {
>         m_resignation = true;
>         m_blackResigned = m_blackToMove;

Because when m_resignation is tested latter to print W+Resign or B+Resign, we're always seeing W+Resign. So maybe it needs to be !m_blackToMove? Assuming black resigns are also reversed, I'm not sure, I'll check database.

>     if(m_resignation) {
>         if (m_blackResigned) {
>             m_winner = QString(QStringLiteral("white"));
>             m_result = "W+Resign ";
>             QTextStream(stdout) << "Score: " << m_result << endl;
>         } else {
>             m_winner = QString(QStringLiteral("black"));
>             m_result = "B+Resign ";
>             QTextStream(stdout) << "Score: " << m_result << endl;
>         }
>

**Comment by isty2e** [comment]
For clarification: This occurs for matches between two different networks (or two leelaz processes), and not for self-plays.

**Comment by roy7** [comment]
The SGF file ends in ";B[tt];W[cp];B[tt];W[qf];B[tt];W[tt]". Google tells me "tt" in SGF means a pass. So actually, white passed after black. White didn't resign per the SGF?

**Comment by isty2e** [comment]
In FastBoard.cpp, line 571–575:

```
    } else if (move == FastBoard::PASS) {
        result << "tt";
    } else if (move == FastBoard::RESIGN) {
	result << "tt";
    }
```

So I assume that was resignation?

**Comment by roy7** [comment]
Hmm. The result in the RE[] SGF field shows a score, not a resignation. Maybe that's normal, I've never studied how this stuff works. SGF file:

> (;GM[1]FF[4]RU[Chinese]DT[2017-12-11]SZ[19]KM[7.5]PB[Leela Zero 0.9 1607dbea]PW[Human]RE[B+17.5]\n\n;B[bm];W[bd];B[gr];W[oq];B[dg];W[bp];B[bg];W[rd];B[mr];W[qq]\n;B[mh];W[rb];B[ck];W[dr];B[bk];W[db];B[rm];W[ld];B[mb];W[og]\n;B[ic];W[lr];B[kb];W[rl];B[ib];W[fr];B[oj];W[pl];B[mj];W[ph]\n;B[gp];W[iq];B[kq];W[dh];B[lq];W[mq];B[ir];W[me];B[nq];W[kr]\n;B[mp];W[nr];B[jq];W[jr];B[or];W[eg];B[ns];W[hn];B[er];W[eq]\n;B[fq];W[es];B[ef];W[df];B[ep];W[cg];B[hq];W[hr];B[ol];W[is]\n;B[pm];W[fp];B[fo];W[rr];B[ip];W[ir];B[ql];W[hp];B[ho];W[qk]\n;B[pk];W[dp];B[do];W[eo];B[en];W[jm];B[rk];W[co];B[gn];W[bh]\n;B[dq];W[cq];B[gb];W[dn];B[dk];W[gh];B[qj];W[eo];B[dm];W[do]\n;B[hm];W[fn];B[fm];W[pr];B[je];W[oe];B[op];W[gj];B[if];W[lg]\n;B[pq];W[lb];B[lc];W[mc];B[la];W[kc];B[lb];W[nb];B[mg];W[kd]\n;B[nc];W[md];B[jg];W[lf];B[lh];W[nd];B[gf];W[fb];B[fc];W[hj]\n;B[fe];W[ec];B[ed];W[gc];B[fd];W[dd];B[gd];W[rh];B[eb];W[ea]\n;B[ri];W[pp];B[de];W[ce];B[qp];W[oq];B[cf];W[rp];B[dg];W[bf]\n;B[ch];W[pq];B[be];W[cd];B[af];W[ci];B[fg];W[cg];B[eh];W[oc]\n;B[ch];W[os];B[di];W[nr];B[mq];W[ms];B[qh];W[jb];B[hc];W[qg]\n;B[rg];W[qi];B[pi];W[qh];B[jc];W[bb];B[lm];W[jo];B[jp];W[cm]\n;B[cl];W[kg];B[kh];W[fp];B[gq];W[jh];B[ih];W[ji];B[ii];W[ij]\n;B[jj];W[ki];B[kj];W[em];B[el];W[li];B[lj];W[hk];B[mi];W[oo]\n;B[cn];W[np];B[ja];W[mo];B[mn];W[lo];B[ln];W[ko];B[in];W[nn]\n;B[nm];W[jn];B[jl];W[im];B[il];W[io];B[on];W[no];B[gs];W[hn]\n;B[hl];W[bn];B[in];W[pn];B[sl];W[hn];B[hp];W[qo];B[in];W[cc]\n;B[an];W[bo];B[qn];W[rn];B[po];W[hn];B[pn];W[ep];B[sn];W[ro]\n;B[in];W[so];B[hn];W[ke];B[jd];W[ar];B[ng];W[oh];B[bi];W[oi]\n;B[cj];W[qd];B[fa];W[ga];B[nk];W[si];B[ha];W[rj];B[fa];W[op]\n;B[eb];W[mm];B[da];W[ca];B[dc];W[ea];B[ml];W[da];B[ad];W[ac]\n;B[ab];W[ec];B[bc];W[fb];B[ac];W[ga];B[ba];W[aa];B[fa];W[ae]\n;B[ad];W[ga];B[sj];W[pj];B[ri];W[am];B[rj];W[cm];B[al];W[sh]\n;B[fa];W[ls];B[eb];W[ac];B[cn];W[ae];B[bf];W[ao];B[fb];W[kf]\n;B[dc];W[jf];B[fs];W[er];B[na];W[pb];B[kp];W[nf];B[lp];W[cm]\n;B[hg];W[am];B[cn];W[kn];B[an];W[cm];B[km];W[am];B[cn];W[br]\n;B[an];W[cm];B[ah];W[am];B[cn];W[ad];B[an];W[cm];B[sm];W[cn]\n;B[ik];W[fi];B[pi];W[oa];B[pj];W[hi];B[jh];W[fk];B[of];W[hh]\n;B[gl];W[am];B[ig];W[bl];B[ej];W[ak];B[aj];W[al];B[ec];W[js]\n;B[ie];W[pf];B[ni];W[fh];B[ma];W[ek];B[fj];W[fl];B[gk];W[mf]\n;B[hs];W[nh];B[qb];W[or];B[rc];W[qc];B[pc];W[rf];B[dl];W[sd]\n;B[om];W[gg];B[od];W[qa];B[ei];W[li];B[gi];W[ob];B[pd];W[pe]\n;B[go];W[id];B[hd];W[qb];B[bj];W[sf];B[id];W[ba];B[ap];W[ds]\n;B[cr];W[pc];B[hh];W[ns];B[cs];W[sb];B[qf];W[sc];B[pd];W[bs]\n;B[qe];W[re];B[cr];W[kl];B[qf];W[sr];B[qr];W[ll];B[mm];W[sg]\n;B[kk];W[qs];B[lk];W[bq];B[ki];W[sq];B[em];W[ss];B[df];W[gj]\n;B[ee];W[ra];B[tt];W[kl];B[tt];W[fi];B[ll];W[fk];B[fh];W[cs]\n;B[nj];W[ek];B[fl];W[fk];B[ek];W[hk];B[tt];W[he];B[hj];W[gh]\n;B[tt];W[hf];B[gg];W[od];B[ge];W[he];B[hf];W[ps];B[tt];W[rs]\n;B[tt];W[aq];B[tt];W[dq];B[tt];W[se];B[tt];W[ap];B[tt];W[rq]\n;B[tt];W[sa];B[tt];W[pa];B[tt];W[an];B[tt];W[bm];B[tt];W[as]\n;B[tt];W[ab];B[tt];W[of];B[tt];W[nc];B[tt];W[qe];B[tt];W[qp]\n;B[tt];W[cp];B[tt];W[qf];B[tt];W[tt])

**Comment by isty2e** [comment]
I thought you were talking about games that ended up with resignation. Seems like tt can be both pass and resignation, at least for leelaz?

**Comment by roy7** [comment]
Yes but I'm unsure if that's right. Per http://www.red-bean.com/sgf/user_guide/#move_vs_place the result of a resignation in the RE[] field should be B+R or B+Resign if black wins. Not B+17.5.

Other resources tell me:

> Since ss is the lower right hand corner of a 19x19 board, tt was used to denote a pass. But larger boards are sometimes used, and a pass should now be written B[] or W[].

As far as I can tell, a resign by the opponent is not recorded as a move in the moves list. The game simply ends and the result is set to RE[B+R] or RE[W+R] depending on winner.

What this SGF file is saying, as recorded, is that white passed after black and black won by 17.5 points.

**Comment by isty2e** [comment]
Oh, that is particularly weird...

**Comment by roy7** [comment]
If someone wants to experiment with this, you could try running a local validation game with resignation percent set to like 75%.

**Comment by roy7** [comment]
Strange, using validation mode even if I set -r 90, it isn't resigning...

**Comment by roy7** [comment]
I'm running fast validation matches looking for a resign. But here was an odd one:

> 340 (W pass) 341 (B pass) 342 (W pass)

**Comment by kimitsu** [comment]
is this
```
   do {
       first.move();
       if(!first.waitForMove()) {
           return res;
       }
       first.readMove();
       second.setMove(bmove + first.getMove());
       second.move();
       if(!second.waitForMove()) {
           return res;
       }
       second.readMove();
       first.setMove(wmove + second.getMove());
       second.nextMove();
   } while (first.nextMove() && m_state.load() == RUNNING);
```
consistent with the idea that the game might end on black's turn?

**Comment by roy7** [comment]
In my super fast validation games, here's an SGF that resulted in 3 passes in a row.

> (;GM[1]FF[4]RU[Chinese]DT[2017-12-10]SZ[19]KM[7.5]PB[Leela Zero 0.9 eebb910d]PW[Human]RE[B+5.5]
> 
> ;B[ic];W[kq];B[kc];W[iq];B[qk];W[hm];B[og];W[gi];B[qi];W[co]
> ;B[rg];W[ll];B[pg];W[lf];B[gb];W[pd];B[fh];W[ed];B[ci];W[gh]
> ;B[bi];W[rd];B[cl];W[nk];B[mb];W[qd];B[pm];W[fl];B[mo];W[qp]
> ;B[eq];W[pq];B[md];W[fq];B[ng];W[ep];B[kg];W[ce];B[dm];W[rp]
> ;B[gd];W[np];B[bm];W[lg];B[cg];W[fg];B[cf];W[od];B[kf];W[de]
> ;B[le];W[oj];B[rm];W[ke];B[mf];W[ac];B[om];W[db];B[go];W[hn]
> ;B[fp];W[fd];B[el];W[er];B[lh];W[gn];B[fr];W[dq];B[gq];W[fo]
> ;B[gp];W[eq];B[he];W[mp];B[gr];W[no];B[eo];W[fn];B[mn];W[fi]
> ;B[nn];W[mm];B[nm];W[nl];B[lj];W[qa];B[en];W[ml];B[ik];W[ao]
> ;B[dj];W[op];B[lp];W[ps];B[ei];W[fm];B[lq];W[lo];B[ln];W[do]
> ;B[ko];W[fk];B[em];W[is];B[ql];W[lk];B[ii];W[eh];B[kk];W[aq]
> ;B[dh];W[mj];B[jl];W[ho];B[ki];W[in];B[hf];W[jo];B[mg];W[hg]
> ;B[mq];W[km];B[lr];W[kn];B[hp];W[ip];B[lo];W[ls];B[nq];W[kr]
> ;B[bf];W[be];B[pr];W[nb];B[oq];W[or];B[pp];W[po];B[nr];W[qr]
> ;B[df];W[lm];B[oo];W[pp];B[ir];W[ks];B[mi];W[on];B[pn];W[qn]
> ;B[oo];W[kp];B[on];W[ol];B[ge];W[dr];B[qm];W[pl];B[jn];W[jm]
> ;B[os];W[hq];B[hr];W[tt];B[pr];W[es];B[eg];W[jr];B[qs];W[fe]
> ;B[ma];W[ef];B[ff];W[dg];B[gf];W[eg];B[nc];W[gg];B[rn];W[oc]
> ;B[qo];W[nd];B[mc];W[ej];B[di];W[ro];B[ar];W[ek];B[sd];W[dk]
> ;B[re];W[dl];B[ck];W[dn];B[cm];W[rs];B[ae];W[hs];B[ob];W[sa]
> ;B[eb];W[gc];B[ia];W[ps];B[so];W[fb];B[fc];W[ec];B[qs];W[ab]
> ;B[rr];W[ps];B[qq];W[rq];B[or];W[qs];B[sq];W[sr];B[ms];W[ad]
> ;B[bq];W[sp];B[dp];W[cp];B[bj];W[af];B[kl];W[se];B[ie];W[fc]
> ;B[qn];W[ae];B[sf];W[sj];B[bp];W[bo];B[na];W[cn];B[br];W[bn]
> ;B[ph];W[sn];B[sh];W[cq];B[se];W[ra];B[ig];W[cr];B[hc];W[bs]
> ;B[bh];W[ai];B[hh];W[hi];B[am];W[fs];B[aj];W[gs];B[ah];W[an]
> ;B[qc];W[sm];B[ih];W[pk];B[kd];W[qj];B[rj];W[rc];B[hj];W[rk]
> ;B[pj];W[rl];B[qj];W[gj];B[sl];W[gk];B[sk];W[qe];B[jj];W[ee]
> ;B[rl];W[so];B[pc];W[oa];B[nb];W[pb];B[lg];W[qb];B[of];W[ag]
> ;B[bc];W[aa];B[bg];W[bd];B[si];W[fa];B[hk];W[ga];B[cj];W[hb]
> ;B[jk];W[hd];B[ib];W[id];B[fp];W[gb];B[ch];W[if];B[jd];W[bb]
> ;B[nj];W[ni];B[im];W[mh];B[nh];W[il];B[hl];W[im];B[gl];W[hd]
> ;B[oi];W[gm];B[nj];W[ok];B[mk];W[cc];B[ap];W[mj];B[sc];W[ea]
> ;B[mk];W[pf];B[oe];W[mj];B[qf];W[cs];B[sb];W[ni];B[pe];W[nj]
> ;B[ba];W[ak];B[ne];W[tt];B[id];W[kh];B[da];W[cb];B[rb];W[ji]
> ;B[jh];W[qc];B[pa];W[ca];B[pc];W[jf];B[cd];W[jg];B[dd];W[dc]
> ;B[je];W[if];B[cd];W[jg];B[jf];W[dd];B[if];W[ja];B[jb];W[nd]
> ;B[jp];W[ld];B[bk];W[lc];B[al];W[lb];B[mr];W[kb];B[la];W[as]
> ;B[ka];W[lc];B[lb];W[jq];B[kb];W[ha];B[ld];W[gr];B[lc];W[od]
> ;B[go];W[gp];B[ir];W[gq];B[mh];W[hr];B[jc];W[qg];B[hd];W[fp]
> ;B[ja];W[qh];B[pi];W[tt];B[tt];W[tt])

**Comment by roy7** [comment]
The other mystery to me is how

> ./leelaz -g  -p 100 --noponder -t 1 -q -d -r 90 -w 1607dbea9c6f8f8e6137b2b167464d4003e5b43c1ec4d792959c6113ef863186

is never resigning. It can't be over 90% all game long...

**Comment by marcocalignano** [comment]
I found the bug with the resignation and **YES** all games won by resignation have the **wrong** winner

**Comment by roy7** [comment]
@kimitsu I think you're on to something. I'm guessing if first.readMove() returns a resign or a 2nd pass in a row, then you need to send to white to end the game properly on white's side for the SGF file, but not ask white for another new move. Asking for a new move after 2 passes is maybe why some games show 3 passes to end.

**Comment by roy7** [comment]
@marcocalignano Would it also affect resigns only showing as White on every game? 

> > db.match_games.find( { score: "W+Resign" } ).count();
> 1964
> > db.match_games.find( { score: "B+Resign" } ).count();
> 0
>

**Comment by killerducky** [comment]
The match request has "resignation_percent": "0" but the command line has "-r 1"?


```
curl -s -J http://zero-test.sjeng.org/get-task/7
190 (W P10) {
    "black_hash": "1607dbea9c6f8f8e6137b2b167464d4003e5b43c1ec4d792959c6113ef863186",
    "cmd": "match",
    "leelaz_version": "0.9",
    "options": {
        "noise": "false",
        "playouts": "1600",
        "randomcnt": "0",
        "resignation_percent": "0"
    },
    "options_hash": "321436",
    "random_seed": "3587476540046543136",
    "required_client_version": "8",
    "white_hash": "b1f25e155577bc849e9447fbe24c20c62094bd23c16e23bd46caf9e509bfcc97"
}

Required client version: 8
 -p 1600  -r 1 -m 0  -t 1  -d   --noponder  -s 3587476540046543136 
Already downloaded network.
Already downloaded network.
./leelaz -p 1600  -r 1 -m 0  -t 1  -d   --noponder  -s 3587476540046543136  -g -q -w 1607dbea9c6f8f8e6137b2b167464d4003e5b43c1ec4d792959c6113ef863186
Engine has started.
Infinite thinking time set.
./leelaz -p 1600  -r 1 -m 0  -t 1  -d   --noponder  -s 3587476540046543136  -g -q -w b1f25e155577bc849e9447fbe24c20c62094bd23c16e23bd46caf9e509bfcc97

```

**Comment by zediir** [comment]
@killerducky are you on the latest commit from next? It's giving me `leelaz.exe -p 1600  -r 0  -m 0  -t 1  -d`

**Comment by killerducky** [comment]
False alarm I guess, I was on a branch but I didn't think it would be a problem. But after going to latest on next it is working correctly. I'll look into what's wrong on my branch.

**Comment by marcocalignano** [comment]
Ok I found also why you got 3 pass and that also could also cause another problem with resignation
Now I need you all to find a way to test it. 
That means find the right option for leelaz to reproduce the problem in gcp next branch 
and run the same option on my next branch to see if it fix (commit is coming)

**Comment by roy7** [comment]
@marcocalignano @kimitsu Maybe marco already made this change, but looking at the code you posted further, not only would 2 passes result in white making a 3rd pass since white is always asked for a move even if game ended on black's turn, but I imagine that's why every Resign is White, since when the game is over as far as autogtp knows, it really is black's turn, since white took a "move after the game ended".

**Comment by marcocalignano** [comment]
@roy7 that is indeed the fix and there is another look at my last commit

**Comment by kimitsu** [comment]
@roy7 While that code explains 3 passes, can't say it easily explains W+Resign, because that part of code should be unaffected, as far as I can see. Let's see what @marcocalignano have come up with.

**Comment by roy7** [comment]
Commit is here https://github.com/marcocalignano/leela-zero/commit/e528a50f544f22972054ff4e79e49a17faccea9a

**Comment by roy7** [comment]
@marcocalignano When trying to do some debugging of this I wanted to force lots of resigned. I changed the validation tool to pass -r 90 to the leela processes. But they still never were resigning. Could you peek at that sometime?

**Comment by marcocalignano** [comment]
@roy7 What did you use for the test?

**Comment by roy7** [comment]
@marcocalignano 

> [roy@ryzen-one validation]$ ./validation  -n 1607dbea9c6f8f8e6137b2b167464d4003e5b43c1ec4d792959c6113ef863186 -n eebb910dc02d50437b8fe78c43e4f77b3428b092e9477163aa16403238cea918 -k sgf
> validation v1
> ./leelaz -g  -p 100 --noponder -t 1 -q -d -r 90 -w eebb910dc02d50437b8fe78c43e4f77b3428b092e9477163aa16403238cea918
> Engine has started.
> Infinite thinking time set.
> ./leelaz -g  -p 100 --noponder -t 1 -q -d -r 90 -w 1607dbea9c6f8f8e6137b2b167464d4003e5b43c1ec4d792959c6113ef863186
> Engine has started.
> Infinite thinking time set.

**Comment by marcocalignano** [comment]
@roy7 I think resignation is inhibited for the first 90 moves

**Comment by roy7** [comment]
It went well past 90 before ending normally. With playouts set to 100 games fly by.

**Comment by killerducky** [comment]
There is also a minimum 500 playouts for resign.

**Comment by roy7** [comment]
> There is also a minimum 500 playouts for resign.

Ahhhh! I never knew that!

**Comment by marcocalignano** [comment]
@roy7 but maybe that is why it never even consider resignation try 500 or 800

**Comment by roy7** [comment]
Well I learned about the 500 playout minimum now, and you've already fixed the bug. So all is well. :)

**Comment by killerducky** [comment]
Here it is, you could change it if you want to test. See also movethresh right above.
https://github.com/gcp/leela-zero/blob/master/src/UCTSearch.cpp#L260

**Comment by roy7** [comment]
Thanks @killerducky I will leave it to the powers that be here https://github.com/gcp/leela-zero/issues/340.

**Comment by marcocalignano** [comment]
@ roy7 can you reproduce the problem?

**Comment by roy7** [comment]
The resign problem? Not particularly. I was just posting the issue for the bug that someone else pointed out in another thread. Hopefully some of the people currently running on /next could pull your commit and try it out. Although I may need to add another match since the current one is about finished.

**Comment by marcocalignano** [comment]
I am testing my fixes 

1 correct:

(B pass) 16 (W pass) Game has ended.

;B[gb];W[cq];B[bg];W[qc];B[rg];W[qa];B[gr];W[cs];B[mr];W[rm]
;B[kq];W[tt];B[ci];W[qk];B[tt];W[tt])

**Comment by marcocalignano** [comment]
2 correct:
 (W resign) Game has ended.
Score: B+Resign 

;B[mk];W[bh];B[pm];W[ng];B[lc];W[ai];B[en];W[ke];B[gp];W[rh]
;B[qg];W[mp];B[mi];W[al];B[hl];W[qe];B[pb];W[tt])

**Comment by roy7** [comment]
I'll set resign % back to normal.

**Comment by marcocalignano** [comment]
@roy7 where?

**Comment by roy7** [comment]
I had set test server to set resignation_percent to 0 on every match while bug was being investigated.

**Comment by marcocalignano** [comment]
And leave it there until the fix is tested and merged then we have to elevate the client version and you have to force the new version before activate resignation again

**Comment by killerducky** [comment]
How about elevate the client version on the test server right now? Then only people with marco's fix can use it.

**Comment by zediir** [comment]
For throughput probably better to just leave it to -r 0 until the fix is merged

**Comment by killerducky** [comment]
I thought there was a pass bug also? Not just resign?

**Comment by marcocalignano** [comment]
3 correct;
(W pass) 77 (B pass) Game has ended.

;B[ip];W[tt];B[gl];W[fm];B[ii];W[tt];B[tt])

**Comment by marcocalignano** [comment]
@killerducky  I did fix them both

**Comment by zediir** [comment]
The pass bug doesn't affect who wins, just add an extra pass to the sgf file so it's not as severe a bug as the resignation one.

**Comment by zediir** [comment]
And the matches don't currently do anything with the passed sgf files right @roy7? Just who won.

**Comment by marcocalignano** [comment]
@zediir but if the third move is not a pass you change the game

**Comment by roy7** [comment]
Correct @zediir.

**Comment by zediir** [comment]
ah right

**Comment by marcocalignano** [comment]
4 correct 
(B resign) Game has ended.
Score: W+Resign 

but the sgf is wrong can anyone tell me what is going on?

;B[im];W[md];B[oo];W[ok];B[hq];W[qk];B[hc];W[kg];B[ee];W[jb]
;B[de];W[od];B[bm];W[kk])

**Comment by killerducky** [comment]
What does the header look like? Leela doesn't output anything at the end for resign, so that part is correct. White played W[kk] and then black resigned. Only the header shows that.

**Comment by marcocalignano** [comment]
yes! there is this in the header!!! 
RE[W+Resign]

**Comment by roy7** [comment]
Correct, resigning is not a "move". So you won't see it in the move list. [] or older [tt] is for passing. For a resign, the game just ends and RE[...] holds the results.

**Comment by roy7** [comment]
Did white really resign though @marcocalignano? Last move was by white, then black didn't make a move. Seems like black resigned?

**Comment by zediir** [comment]
Black resigned. white won by resignation

**Comment by roy7** [comment]
Right sorry... I need sleep...

**Comment by marcocalignano** [comment]
But I think there is problem here: 

the validation program says:
```
(W resign) Game has ended.
Score: B+Resign 
Winner: black

```
but the sgf is this 

```
;GM[1]FF[4]RU[Chinese]DT[2017-12-10]SZ[19]KM[7.5]PB[Leela Zero 0.9 22373747]PW[Human]RE[B+10.5]

;B[bg];W[nq];B[mb];W[fc];B[mr];W[nc];B[gb];W[cf];B[kk];W[ri]
;B[bi];W[tt];B[ic];W[tt];B[dg];W[qk];B[rk];W[bm];B[gr];W[tt]
;B[rm];W[tt];B[nb];W[ac];B[ib];W[qi];B[qp];W[bn];B[pm];W[oc]
;B[ob];W[jh];B[kr];W[lo];B[ci];W[nl];B[od];W[hf];B[nd];W[oe]
;B[op];W[il];B[mp];W[ap];B[ir];W[de];B[ed];W[me];B[hm];W[ae]
;B[ho];W[mh];B[ne];W[km];B[mi];W[ls];B[gd];W[ma];B[ql];W[hj]
;B[if];W[rg];B[nk];W[gj];B[el];W[jl];B[ep];W[qh];B[ek];W[aa]
;B[cl];W[hb];B[ni];W[hn];B[or];W[qf];B[oj];W[ol];B[po];W[aq]
;B[gf];W[cj];B[lc];W[es];B[hg];W[pn];B[qd];W[bq];B[he];W[mq]
;B[fe];W[qo];B[og];W[oh];B[md];W[sr];B[qn];W[go];B[qj];W[jd]
;B[gn];W[bk];B[in];W[pc];B[lj];W[lb];B[gk];W[kf];B[fi];W[ml]
;B[pk];W[pa];B[nh];W[kp];B[fh];W[jb];B[gp];W[lg];B[jr];W[ps]
;B[oa];W[ng];B[hq];W[gg];B[nf];W[ia];B[mg];W[gi];B[fg];W[hi]
;B[gh];W[cg];B[lh];W[js];B[cc];W[pg];B[hl];W[hp];B[np];W[ch]
;B[lp];W[tt])

```
here happened the following:
W resign so The move is read is given to the black process white the gtp command "white resign" the sgf  write a pass "W[tt]"  and then i ask get score and the score is calculated RE[B+10.5] 

Anyona has an idea?

**Comment by roy7** [comment]
If white resigned W[tt] should not get written. [tt] means a pass. Nothing gets written to the SGF file in the move section from resigning.

**Comment by marcocalignano** [comment]
That what I thought but 

```
 if (command.find("play") == 0) {
        if (command.find("pass") != std::string::npos
            || command.find("resign") != std::string::npos) {
            game.play_pass();
            gtp_printf(id, "");
        } else {
```

**Comment by marcocalignano** [comment]
So I my fix is tested and good but I think there is an error on the GTP protocol if you send the command 
"play white resign" then the sgf is wrong!

**Comment by marcocalignano** [comment]
will be fix with #339

**Comment by killerducky** [comment]
Maybe we should redo b1f25e15 match? It was corrupted right? But either disable resign or bump the client versions because some people might be running the buggy code?

**Comment by roy7** [comment]
My plan is to see what @gcp decides to tomorrow, I assume we'll get the PR pulled in and an updated required client version for /next. At that point I can see what I can do. I might be able to just correct the scores manually for the bad games, or maybe just delete the games and update the match results (backing out the wins/losses related to W+Resign) and those games can be replayed.

**Comment by gcp** [comment]
Internally in Leela resigning and passing are encoded and treated rather similarly. In anything but Trump Taylor this makes no difference, but unfortunately...

**Comment by roy7** [comment]
All bad resignations were deleted. (And good ones too, too hard to try and fix it, but wasn't too bad to manually remove all the W+Resign and adjust the matches summary table by hand for it all.)

@marcocalignano I did increase options.required_client_version to "10", but I didn't change the required version that comes back from /best-network-hash since I didn't want to break the main network if I copied test server updates over to live server. I'm curious, does /test demand an update based only off the get-task options, or from /best-network-hash, or either one?

**Comment by marcocalignano** [comment]
The new client V9 and v10 for the next Branch does not even look at  /best-network-hash It download the network hash from the json that is gets so no problem there.
@roy7 If you are confident we could also move all client with version 10 to the main server since the test seems to go good.

**Comment by roy7** [comment]
Well release version doesn't use get-task yet, and /next client doesn't connect to main server. So it's really just a matter of rolling /next out to everyone then updating main server when @gcp is ready. I see some bug reports people are concerned about short games when multiple games play at once, so we want to be sure there's no bugs left of course.

**Comment by marcocalignano** [comment]
@roy7 my client just started a match between these two (real old) networks **ad8a1d02 VS 6f274ab1**

**Comment by roy7** [comment]
Yes I removed all bad resignations and am going to fill back in any of the matches that went back under 400. You can see it on the matches table. :)

**Comment by roy7** [comment]
@isty2e Best network is not automatically updated. It will be automated sometime soon.

**Comment by marcocalignano** [comment]
workaround is in #353

**Comment by Hersmunch** [comment]
Just spotted this recent match http://zero.sjeng.org/viewmatch/9d41e6590178bfbcd31f59c80102d4753d7a5b40d7fb088b26d8863121d37bfa?viewer=eidogo where white appears to resign but it is recorded as RE[B+9.5] in the sgf. (Also, it has a nice opening :) )
@marcocalignano, I see you picked up this earlier but I can't tell from the comments here whether it is thought to have been fixed already or not.

**Comment by roy7** [comment]
I think all the SGF issues have been resolved in the /next branch, but most self play still comes from /release. :)

**Comment by Hersmunch** [comment]
Sorry, I should have made it clearer that it was a match game Leela Zero ce1aa643 vs Leela Zero 0.9 154eaebd

**Comment by gcp** [comment]
If I fix this in the engine now, autogtp won't be confused, right?

**Comment by marcocalignano** [comment]
probably not but I am not really sure. Just fix it and give it to me to test.

**Comment by gcp** [comment]
This should be fixed now with 0.10 release.
